### ISSUE TITLE
Migration 0022_dbgroup_type_string_change_content (Django) takes too long

### ISSUE THREAD

**Issue by szoupanos** [aiidateam/aiida_core#2797]
This migration very likely is wrong.
I started to look at it since it takes **more than 2 days** on a 7M node database

According to the description we would like to migrate CalcJobNode attributes for metadata options whose key changed.

    Renamed attribute keys:

      * `custom_environment_variables` -> `environment_variables`
      * `jobresource_params` -> `resources`
      * `_process_label` -> `process_label`
      * `parser` -> `parser_name`

    Deleted attributes:
      * `linkname_retrieved` (We do not actually delete it just in case some relies on it

If I have a look at the first update sql query, I don't see any join between the attributes and the nodes:
```
UPDATE db_dbattribute AS attribute
SET key =  regexp_replace(attribute.key, '^custom_environment_variables', 'environment_variables')
FROM db_dbnode AS node
WHERE
	(
		attribute.key = 'custom_environment_variables' OR
		attribute.key LIKE 'custom\_environment\_variables.%'
	) AND
	node.type = 'node.process.calculation.calcjob.CalcJobNode.';
```

I analyze the execution plan of the query (to ensure that no join will be added - e.g. implicitly) and I see the following:

```
                                                                          QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------
 Update on db_dbattribute attribute  (cost=1856.29..14419903.13 rows=810381824 width=569)
   ->  Nested Loop  (cost=1856.29..14419903.13 rows=810381824 width=569)
         ->  Bitmap Heap Scan on db_dbattribute attribute  (cost=1510.70..205773.67 rows=77504 width=63)
               Recheck Cond: (((key)::text = 'custom_environment_variables'::text) OR ((key)::text ~~ 'custom\_environment\_variables.%'::text))
               Filter: (((key)::text = 'custom_environment_variables'::text) OR ((key)::text ~~ 'custom\_environment\_variables.%'::text))
               ->  BitmapOr  (cost=1510.70..1510.70 rows=57439 width=0)
                     ->  Bitmap Index Scan on db_dbattribute_key_6936ff5c4f96a1be_like  (cost=0.00..1467.37 rows=57439 width=0)
                           Index Cond: ((key)::text = 'custom_environment_variables'::text)
                     ->  Bitmap Index Scan on db_dbattribute_key_6936ff5c4f96a1be_like  (cost=0.00..4.58 rows=1 width=0)
                           Index Cond: (((key)::text ~>=~ 'custom_environment_variables.'::text) AND ((key)::text ~<~ 'custom_environment_variables/'::text))
         ->  Materialize  (cost=345.59..32473.68 rows=10456 width=6)
               ->  Bitmap Heap Scan on db_dbnode node  (cost=345.59..32421.40 rows=10456 width=6)
                     Recheck Cond: ((type)::text = 'node.process.calculation.calcjob.CalcJobNode.'::text)
                     ->  Bitmap Index Scan on db_dbnode_type_4cda33f938ccd765_like  (cost=0.00..342.98 rows=10456 width=0)
                           Index Cond: ((type)::text = 'node.process.calculation.calcjob.CalcJobNode.'::text)
```

Which mentions at the top level, that it results to 810.381.824 rows (!!!) when the attribute table has "only" 282.505.715. So it seems to apply the attribute update to every attribute that satisfies the condition (irrespective to which type of node it corresponds to) and this is repeated for all the nodes that satisfy the condition.

I didn't check if it just runs the update db_dbnode times (for the calc nodes) or if it adds db_dbnode times (for the calc nodes) updated attributes to the table. But this is less important.

After adding the join condition to the update
```
UPDATE db_dbattribute AS attribute
SET key =  regexp_replace(attribute.key, '^custom_environment_variables', 'environment_variables')
FROM db_dbnode AS node
WHERE
	(
		attribute.key = 'custom_environment_variables' OR
		attribute.key LIKE 'custom\_environment\_variables.%'
	) 
	AND
	node.type = 'node.process.calculation.calcjob.CalcJobNode.'
	AND
	node.id = attribute.dbnode_id;
```

I get the following execution plan which is more normal (in the number of expected rows)
```
                                                                          QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------
 Update on db_dbattribute attribute  (cost=34062.80..238618.06 rows=110 width=569)
   ->  Hash Join  (cost=34062.80..238618.06 rows=110 width=569)
         Hash Cond: (attribute.dbnode_id = node.id)
         ->  Bitmap Heap Scan on db_dbattribute attribute  (cost=1510.70..205773.67 rows=77504 width=63)
               Recheck Cond: (((key)::text = 'custom_environment_variables'::text) OR ((key)::text ~~ 'custom\_environment\_variables.%'::text))
               Filter: (((key)::text = 'custom_environment_variables'::text) OR ((key)::text ~~ 'custom\_environment\_variables.%'::text))
               ->  BitmapOr  (cost=1510.70..1510.70 rows=57439 width=0)
                     ->  Bitmap Index Scan on db_dbattribute_key_6936ff5c4f96a1be_like  (cost=0.00..1467.37 rows=57439 width=0)
                           Index Cond: ((key)::text = 'custom_environment_variables'::text)
                     ->  Bitmap Index Scan on db_dbattribute_key_6936ff5c4f96a1be_like  (cost=0.00..4.58 rows=1 width=0)
                           Index Cond: (((key)::text ~>=~ 'custom_environment_variables.'::text) AND ((key)::text ~<~ 'custom_environment_variables/'::text))
         ->  Hash  (cost=32421.40..32421.40 rows=10456 width=10)
               ->  Bitmap Heap Scan on db_dbnode node  (cost=345.59..32421.40 rows=10456 width=10)
                     Recheck Cond: ((type)::text = 'node.process.calculation.calcjob.CalcJobNode.'::text)
                     ->  Bitmap Index Scan on db_dbnode_type_4cda33f938ccd765_like  (cost=0.00..342.98 rows=10456 width=0)
                           Index Cond: ((type)::text = 'node.process.calculation.calcjob.CalcJobNode.'::text)
```

I will update the migration queries and I report again how it goes.
We should add also 1 test that demonstrates that attributes of other nodes are nodes are not updated

**Comment by szoupanos** [comment]
So there is a PR for this (#2798). What I noticed is that attributes.resources is a string and not a dictionary (jobresource_params was a dictionary).

@sphuber expects this since the test in SQLA is written explicitly to expect a string and not a dictionary even if attributes.jobresource_params was a dictionary

The same should stand for custom_environment_variables/environment_variables.

@sphuber @giovannipizzi 
Is this something that we want (to store values of a key that is a dictionary as a string)?

**Comment by sphuber** [comment]
@szoupanos I have updated PR #2798 to address also the bug in the SqlAlchemy migration. We absolutely do *not* want to convert dictionaries to strings. This was just a monumental fuck up by yours truly.

What I noticed here however, is that the title of this issue references the migration *before* the one you then describe. I take it the title is wrong and it is migration `1.0.23` and not `1.0.22` that is slow, correct? Would that be fixed by the correct join in the Django migration, that will be applied after merging PR #2798?

**Comment by szoupanos** [comment]
> What I noticed here however, is that the title of this issue references the migration _before_ the one you then describe. I take it the title is wrong and it is migration `1.0.23` and not `1.0.22` that is slow, correct? Would that be fixed by the correct join in the Django migration, that will be applied after merging PR #2798?

@sphuber Yes, you are right. The title of this issue is wrong (it should mention migration 0.23 and not 0.22) and the fix of PR #2798 should fix the problem described here. I will update the title now
